File: code/quicksort/cpp/2105.cpp
#include<iostream>
using namespace std;
int partition(int arr[],int l,int r)
{
  int i=l-1,temp;
  for(int j=l;j<r;j++)
    if(arr[j]>arr[r])
    {
      i++;
      temp=arr[i];
      arr[i]=arr[j];
      arr[j]=temp;
    }
  temp=arr[r];
  arr[r]=arr[++i];
  arr[i]=temp;
  return i;
}
void QuickSort(int arr[],int low,int high)
{
  if(low<high)
  {
    int p=partition(arr,low,high);
    QuickSort(arr,low,p-1);
    QuickSort(arr,p+1,high);
  }
}
int main()
{
 int N;cin>>N;
 while(N--)
 {
   int n;cin>>n;
   int arr[n];
   for(int i=0;i<n;i++)cin>>arr[i];
   QuickSort(arr,0,n-1);
   for(int i=0;i<n;i++)cout<<" "<<arr[i];
   cout<<endl;
}
}

File: code/quicksort/java/2105.java
public class Quicksort {

	public static void quickSort(int[] array) {
		sort(array, 0, array.length-1);
	}

	public static void sort(int[] array, int low, int high) {
		if(high > low) {
			int pivot = partition(array, low, high);
			sort(array, low, pivot-1);
			sort(array, pivot+1, high);
		}
	}

	public static int partition(int[] array, int low, int high) {
		int pivot = array[high];
		int i = low-1;
		
		for(int j = low; j < high; j++) {
			if(array[j] <= pivot) {
				i++;
				swap(array, i, j);
			}
		}

		swap(array, i+1, high);
		return i+1;
	}

	public static void swap(int[] array, int i, int j) {
		int temp = array[i];
		array[i] = array[j];
		array[j] = temp;
	}

	public static void main(String args[]) {
		int[] array = {3,5,2,5,2,4,7,9,65,2,1};
		quickSort(array);

		for(int i = 0; i < array.length; i++) {
			System.out.print(array[i] + " ");
		}
	}
}
